Documentation/design/assetgeneration.md

```diff
@@ -0,0 +1,124 @@
+# Installer asset generation based on dependency resolution
+
+## Goal
+
+Define generation of assets of varied types in installer based on a dependency graph.
+
+## Overview
+
+The installer generates assets based on the dependency [graph](./dependency.md). To generate an asset, all its dependencies have to be generated. The installer needs to define how assets declare this dependency and how it resolves the graph when asked to generated specific targets. Targets are chosen assets that need to be generated, the installer generates all assets that the target depends on and writes to disk only the targets, consuming any inputs from disk or state from a previous run.
+
+Each asset is individually responsible for declaring its dependencies. Each asset is also responsible resolving conflicts when combining its input from disk and its state from a previous run. The installer ensures all the dependencies for an asset is generated and provides the asset with latest state to generate its own output.
+
+## State
+
+State is an internal state of the installer. It stores the contents of all the assets that have been generated.
+
+An example of the State:
+
+```go
+// State is map of assets
+type State struct {
+    Root string // hash for state object.
+
+    Objects map[string]AssetState
+}
+
+// AssetState is the state of an Asset.
+type AssetState struct {
+    Parents maps[string]string // hashes for our each parent.
```

Comment:
Why did this end up being a map?  I think it should be a slice of hash strings; see [here][1] and [here][2].

[1]: https://github.com/openshift/installer/pull/105#discussion_r207412024
[2]: https://github.com/openshift/installer/pull/105#discussion_r207695151